libs/clients/zebpay/client.go

```diff
@@ -0,0 +1,297 @@
+type ClientOpts struct {
+	APIKey     string
+	SigningKey crypto.PrivateKey
+}
+
+// Client abstracts over the underlying client
+type Client interface {
+	// BulkTransfer posts a bulk transfer request to zebpay
+	BulkTransfer(ctx context.Context, opts *ClientOpts, req BulkTransferRequest) (*BulkTransferResponse, error)
+	// BulkCheckTransfer checks the status of a transaction
+	BulkCheckTransfer(ctx context.Context, opts *ClientOpts, ids ...uuid.UUID) (BulkCheckTransferResponse, error)
+	// CheckTransfer checks the status of a transaction
+	CheckTransfer(ctx context.Context, opts *ClientOpts, id uuid.UUID) (*CheckTransferResponse, error)
+}
+
+// HTTPClient wraps http.Client for interacting with the cbr server
+type HTTPClient struct {
+	client *clients.SimpleHTTPClient
+}
+
+// New returns a new HTTPClient, retrieving the base URL from the environment
+func New() (Client, error) {
+	serverEnvKey := "ZEBPAY_SERVER"
+	serverURL := os.Getenv(serverEnvKey)
+	if len(serverURL) == 0 {
+		return nil, errors.New(serverEnvKey + " was empty")
+	}
+	proxy := os.Getenv("HTTP_PROXY")
+	client, err := clients.NewWithProxy("zebpay", serverURL, "", proxy) // authentication bearer token is set per api call
+	if err != nil {
+		return nil, err
+	}
+	return NewClientWithPrometheus(&HTTPClient{client}, "zebpay_client"), err
+}
+
+var errBadClientOpts = errors.New("client is misconfigured with no client options")
+
+type claims struct {
+	*jwt.Claims           // use iat, exp, sub - same as x-api-key
+	URI         string    `json:"uri,omitempty"`      // the endpoint called
+	Nonce       uuid.UUID `json:"nonce,omitempty"`    // one time use uuid
+	BodyHash    string    `json:"bodyHash,omitempty"` //Hex-encoded SHA-256 hash of the raw HTTP request body.
+}
+
+// affixAccessToken will take a request and generate a zebpay access token and affix it to the
+// headers of the request
+func affixAccessToken(req *http.Request, opts *ClientOpts, body []byte) error {
+	if opts == nil {
+		return errBadClientOpts
+	}
+
+	// sha body bytes
+	h := sha256.New()
+	h.Write(body)
+	shasum := fmt.Sprintf("%x", h.Sum(nil))
+
+	// add the api key to the request
+	req.Header.Set("X-API-KEY", opts.APIKey)
+
+	// setup the signer options
+	key := jose.SigningKey{Algorithm: jose.RS256, Key: opts.SigningKey}
+	var signerOpts = jose.SignerOptions{}
+	signerOpts.WithType("JWT")
+
+	// create a new signer with the signing key
+	rsaSigner, err := jose.NewSigner(key, &signerOpts)
+	if err != nil {
+		return fmt.Errorf("failed to create jwt signer: %w", err)
+	}
+
+	// attach the rsa signer to the builder
+	builder := jwt.Signed(rsaSigner)
+
+	// build out the claims of the token
+	builder = builder.Claims(claims{
+		Claims: &jwt.Claims{
+			Subject:  opts.APIKey,
+			IssuedAt: jwt.NewNumericDate(time.Now()),
+			Expiry:   jwt.NewNumericDate(time.Now().Add(time.Hour)),
+		},
+		URI:      req.URL.Path,
+		Nonce:    uuid.New(),
```

Comment:
Is this uuid v4 by default?

Reply:
yes uuidv4 default

Reply:
thanks for confirming - I was on mobile and couldn't tell right away